File: day1/day1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func puzzle1(lines []string, part string) {
	total := 0
	matcher, _ := regexp.Compile("\\d")
	for _, line := range lines {
		matches := matcher.FindAllString(line, -1)
		value := matches[0] + matches[len(matches)-1]
		intVal, _ := strconv.Atoi(value)
		total = total + intVal
	}
	fmt.Println("Answer for part", part, total)
}

func puzzle2(lines []string) {
	newLines := []string{}

	replaceMap := make(map[string]string)
	replaceMap["one"] = "o1e"
	replaceMap["two"] = "t2o"
	replaceMap["three"] = "t3e"
	replaceMap["four"] = "f4r"
	replaceMap["five"] = "f5e"
	replaceMap["six"] = "s6x"
	replaceMap["seven"] = "s7n"
	replaceMap["eight"] = "e8t"
	replaceMap["nine"] = "n9e"

	for _, line := range lines {
		newLine := line
		for old, new := range replaceMap {
			newLine = strings.ReplaceAll(newLine, old, new)
		}
		newLines = append(newLines, newLine)
	}

	puzzle1(newLines, "2")
}

func main() {
	file, err := os.Open("../Inputs/day1.txt")
	if err != nil {
		fmt.Printf("error opening file: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	lines := []string{}
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		fmt.Printf("error processing file: %v\n", err)
		os.Exit(1)
	}

	puzzle1(lines, "1")
	puzzle2(lines)

}

File: day1/day1.py
file1 = open('../Inputs/day1.txt', 'r')
lines = file1.readlines()
total = 0

word_replacements = {
    'one': '1',
    'two': '2',
    'three': '3',
    'four': '4',
    'five': '5',
    'six': '6',
    'seven': '7',
    'eight': '8',
    'nine' : '9',
    'zero': '0'
}

part1_total = 0
part2_total = 0

for line in lines:
    l = line.strip()
    
    simple_nums = [c for c in l if c.isdigit()]
    val = simple_nums[0]
    val = val + simple_nums[-1]
    part1_total = part1_total + int(val)

    ix_nums = dict([c for c in enumerate(l) if c[1].isdigit()])
    for i in range(len(line)):
        for w in word_replacements:
            if l.find(w, i, i + len(w)) != -1:
                ix_nums[i] = word_replacements[w]
    sorted_ix_nums = dict(sorted(ix_nums.items()))
    nums = [ix_nums[c] for c in sorted_ix_nums]

    val = nums[0]
    val = val + nums[-1]
    part2_total = part2_total + int(val)

print("Part 1:", part1_total)
print("Part 2:", part2_total)